data-processing/scripts/compute_accuracy.py

```diff
@@ -68,20 +69,54 @@ def fetch_boxes(
             BoundingBoxModel.width,
             BoundingBoxModel.height,
             BoundingBoxModel.page,
+            # Aggregate data for an entity into an array, where each field
+            # is a dictionary: {"key", "...", "value", "..."}. All values will
+            # be of type string.
+            fn.json_agg(
+                fn.json_build_object(
+                    "key", EntityDataModel.key, "value", EntityDataModel.value
+                )
+            ).alias("data"),
         )
         .join(Paper)
         .switch(EntityModel)
         .join(BoundingBoxModel)
+        .switch(EntityModel)
+        .join(EntityDataModel)
         .where(
             EntityModel.version == version,
             Paper.arxiv_id == arxiv_id,
             EntityModel.type << types,
         )
+        .group_by(
+            EntityModel.id,
+            EntityModel.type,
+            BoundingBoxModel.left,
+            BoundingBoxModel.top,
+            BoundingBoxModel.width,
+            BoundingBoxModel.height,
+            BoundingBoxModel.page,
+        )
         .dicts()
     )
+
     boxes_by_entity_db_id: Dict[str, List[BoundingBox]] = defaultdict(list)
     types_by_entity_db_id: Dict[str, str] = {}
     for row in rows:
+
+        # When processing symbols, ignore those that are not of the types included in the gold set.
+        # For example, operators will be ignored because they are not in the gold set. If a symbol
+        # does not have an explicit type, it is assumed to be of one of the allowed types.
+        GOLD_SYMBOL_TYPES = ["function", "identifier"]
```

Comment:
Not a big deal, but any reason this filter shouldn't be in the query above as a where clause instead?

Reply:
Yup! I was thinking of putting it in the query above. The query was becoming pretty complex (as the desired behavior was to:

1. outer join on the entitydata table on columns where key is 'type', but only for symbols
2. if it is a symbol and it has type but is not one of the gold types, filter it out
3. for all other entity types (e.g., sentences, citations, etc.) do not impose this constraint

I thought it might be more readable without a noticeable hit to performance if I did a simple aggregation of keys and values, returned one row per entity, and did the nuanced conditionals in the Python code. Any thoughts on whether it makes sense to move into the query instead?

Reply:
If the query would be much more gross than this logic I'm down with keeping it as is, I just wasn't sure if there was a way for it to be about the same in query form.